### ISSUE TITLE
Weird Duplication of Characters in Names

### ISSUE THREAD

**Issue by SGifto73** [AtomCrafty/UniversalInjectorFramework#1]
Hello, I'm not sure if this project is still getting active development but I figured I'd shoot my shot anyways.

I'm using this tool for the translation of a visual novel using the Yu-RIS game engine (https://vndb.org/v5928), and it almost works perfectly, except for one small issue.

![image](https://user-images.githubusercontent.com/125715965/220776715-81e836e1-851b-48b5-ab06-9d28e4e6a7db.png)
For some reason, characters in names are duplicated, but only English names. The protagonist is renamable in this game but it isn't just her that suffers from the bug.

![image](https://user-images.githubusercontent.com/125715965/220776970-ece3f2ab-e2ee-4b33-93f0-f699ad7e61e1.png)
Another name suffering from the duplication glitch.

![image](https://user-images.githubusercontent.com/125715965/220777090-9712d9ed-88e5-41a4-b3bc-ac60b772eabe.png)
As you can see though, if I leave her name in Japanese, the duplication doesn't occur.

I narrowed down the bug to the "character replace" feature specifically, the problem is I need to use that to get rid of the "Japanese quotation marks" (「」) as these are generated by the engine I'm pretty sure (I don't see them in the script.)

So I guess my question is, why would the character replace function cause this? And why would it duplicate on the second characters and onward in each name?

[Files.zip](https://github.com/AtomCrafty/UniversalInjectorFramework/files/10808904/Files.zip)
Here's the DLL I'm using along with the config file.

**Comment by AtomCrafty** [comment]
Hmm, I'm not quite sure what could be causing this.
Do you know which text API the engine uses (`TextOut` / `GetGlyphOutline`)?
If not, could you upload the engine .exe for me so I can check?

**Comment by SGifto73** [comment]
![image](https://user-images.githubusercontent.com/125715965/220784143-1376f515-9954-4557-9999-3888bc724165.png)
I should've screenshotted the console file too, sorry about that. Seems like it uses TextOut.

**Comment by SGifto73** [comment]
![image](https://user-images.githubusercontent.com/125715965/220785449-361c696a-b525-4b01-8bf0-1da67c3ac4cb.png)
I figured I could also show you what happens with character substitution turned off. As you can see the name renders fine, but that "u" and "v" are 「 and 」respectively, so I need to get rid of them with the character substitution tool.

**Comment by AtomCrafty** [comment]
[winmm.zip](https://github.com/AtomCrafty/UniversalInjectorFramework/files/10809264/winmm.zip)
Please try running it with this version of the .dll. I just added a line to dump all strings to the console before and after substitution.

**Comment by SGifto73** [comment]
![image](https://user-images.githubusercontent.com/125715965/220792425-0832a2d2-34a0-4c8c-aed5-d3e95bdc09c9.png)
![image](https://user-images.githubusercontent.com/125715965/220792502-f57d9f94-5dfe-4d5b-94be-a1a7f7956c6b.png)
![image](https://user-images.githubusercontent.com/125715965/220792538-f94380b1-fdc8-4e4f-8f24-036ef832527a.png)

Here's the results, every character seems to be in there 5 times, most of it has no anomalies other than that so I tried to just capture the parts with anything weird possibly happening.

**Comment by SGifto73** [comment]
Probably easier to put it in a text file.
[test.txt](https://github.com/AtomCrafty/UniversalInjectorFramework/files/10809295/test.txt)

**Comment by AtomCrafty** [comment]
Okay, I believe I understand what's going on.
The engine tries to draw the name one character at a time, so for the name 伊藤綾子 it would call `TextOut` four times, each time passing 1 for the `c` parameter to only draw one character.

```
伊
　藤
　　綾
　　　子

伊藤綾子 ⟵ final result
```

My `TextOut` hook however ignores the `c` parameter, so it draws the full string every time.
With the Japanese name that isn't immediately obvious, because even though 子 is drawn multiple times, it is drawn in (almost) the same location each time.

```
伊藤綾子
　藤綾子
　　綾子
　　　子

伊藤綾子 ⟵ final result
```

The alignment isn't perfect though, you can actually see some remnants of the first 子 in the screenshot you sent earlier.

![image](https://user-images.githubusercontent.com/4553050/220799453-8cf72fa1-0c45-467c-a1fb-b6dfe4f9d6a3.png)

Now, where that issue really becomes obvious is when we start using latin characters. They are much narrower than the full-width kanji, but the engine still moves the cursor by the same number of pixels it would expect a full width character to take up.

```
Ayako
  yako
    ako
      ko
        o

Ayyaakkoo ⟵ final result
```

I will fix the hook to properly handle the `c` parameter, but I can't do much about the fact that the letters will be very far apart, as that is caused by the engine itself.

```
A
  y
    a
      k
        o

A y a k o ⟵ final result
```

I guess you could add some custom code that ignores the `c` parameter if and only if the string passed to it is exactly one of the character names, and ignore all subsequent `TextOut` calls with suffixes of said name.

```
Ayako      ⟵ matches a known name, so draw it entirely
 (yako)    ⟵ suffix of a name, not drawn at all
   (ako)   ⟵ suffix of a name, not drawn at all
     (ko)  ⟵ suffix of a name, not drawn at all
       (o) ⟵ suffix of a name, not drawn at all

Ayako      ⟵ final result
```

**Comment by SGifto73** [comment]
Thanks a lot for the response! yeah I see the first 子 in the screenshot you posted, I see how it works now. The problem is that because the protagonist is nameable, I don't know if I could pull the "known name", unless there's a way to somehow fetch it, so that leaves me I think with either using the wide names after you fix that or just telling end users to use the default name (which I ideally wanted to avoid but I wouldn't be entirely against it), thank you a lot though.

**Comment by AtomCrafty** [comment]
Fixed by f4d3ab2

**Comment by AtomCrafty** [comment]
> The problem is that because the protagonist is nameable, I don't know if I could pull the "known name", unless there's a way to somehow fetch it, so that leaves me I think with either using the wide names after you fix that or just telling end users to use the default name (which I ideally wanted to avoid but I wouldn't be entirely against it), thank you a lot though.

I have a few ideas how to work around it without having to hard code the names, but I believe that discussion would be more suited for Discord than a GitHub issue. My Discord tag is AtomCrafty#4511, if that's fine with you.
